✨  Add support for generic components using `FieldPathWithValue`

julienfouilhe opened the pull request:
This seems like a better and much simpler API for the support of generic components than #6466

If the expected type is not specified to useController/`Controller`, then all possible types from the `FieldValues` are returned.

You can look at the test in `useController.test.tsx` for how it can be used.

julienfouilhe committed:
✨  Add support for generic components using FieldPathWithValue

julienfouilhe committed:
✅  Fix tests

julienfouilhe commented:
cc @bluebill1049 @kotarella1110 and @barrymay since you were tagged as reviewers on my other PR.

I could not find a way to keep the tests the same, I had to explicitly pass the types to `userController`/`Controller`. I think there may be a way to infer everything but I couldn't find a way.

If someone better at Typescript than me knows of a way, please tell me 🙏

julienfouilhe committed:
✨  Fix inference

julienfouilhe committed:
Revert "✅  Fix tests"

This reverts commit 2058b5b29fc6d9190a6bcbcdd9e59e1d66c861c3.

julienfouilhe committed:
🎨  Improve readibility of `FieldPathWithValue`

julienfouilhe commented on src/useController.ts:
```diff
@@ -17,10 +17,14 @@ import { useFormState } from './useFormState';
 export function useController<
   TFieldValues extends FieldValues = FieldValues,
-  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
+  TResult = any,
```
This needs to be `any` if unspecified because we want `TName` to be able to be one of all the possible field paths of `TFieldValues`.
There's a check where it's actually used so TResult should never really be used as `any`

barrymay replied:
Superb work - thanks for the explanation.  I look forward to leveraging this!

julienfouilhe commented on src/types/controller.ts:
```diff
@@ -24,49 +25,69 @@ export type ControllerFieldState<
   >;
   shouldUnregister?: boolean;
-  defaultValue?: UnpackNestedValue<FieldPathValue<TFieldValues, TName>>;
+  defaultValue?: IsAny<TResult> extends true
+    ? FieldPathValue<TFieldValues, TName>
+    : TResult;
```
If the expected type is `any`, we try to resolve the type of the field using the field values type and the path.

If it's specified, we use that type.

That allows for components using:

```typescript
type FormValues = { test: string; test: string[] };

const { control } = useForm<FormValues>;

useController({ control, name: 'test', defaultValue: 'str' });
```
to resolve `TResult` to `FormValues[name]` because it knows both the FormValues and the name key.

But it also allows generic components, for which we do not know the FormValues and the path, to actually fix the expected type.

bluebill1049 replied:
Thank you so much for this amazing contribution ❤️

bluebill1049 reviewed (dismissed):
LGTM.

bluebill1049 commented:
let's wait for @barrymay and @kotarella1110's review feedback 🙏

bluebill1049 committed:
Merge branch 'master' into generic-components

kotarella1110 commented on src/types/controller.ts:
```diff
@@ -24,49 +25,69 @@ export type ControllerFieldState<
 
 export type ControllerRenderProps<
   TFieldValues extends FieldValues = FieldValues,
-  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
+  TResult = any,
+  TName extends FieldPathWithValue<TFieldValues, TResult> = FieldPathWithValue<
+    TFieldValues,
+    TResult
+  >,
 > = {
   onChange: (...event: any[]) => void;
   onBlur: () => void;
-  value: UnpackNestedValue<FieldPathValue<TFieldValues, TName>>;
+  value: IsAny<TResult> extends true
+    ? FieldPathValue<TFieldValues, TName>
+    : TResult;
   name: TName;
   ref: RefCallBack;
 };
```
Does `ControllerRenderProps` and `UseControllerProps` type need `TResult` ?
I think that `TResult` can be infer from `TName` 👀 

As follows: 

```ts
export type ControllerRenderProps<
  TFieldValues extends FieldValues = FieldValues,
  TName extends FieldPathWithValue<TFieldValues, TResult> = FieldPathWithValue<
    TFieldValues
  >,
> = {
  onChange: (...event: any[]) => void;
  onBlur: () => void;
  value: TName extends FieldPathWithValue<TFieldValues, infer TResult>
    ? IsAny<TResult> extends true
      ? FieldPathValue<TFieldValues, TName>
      : TResult
  : FieldPathValue<TFieldValues, TName>;
  name: TName;
  ref: RefCallBack;
};
```

julienfouilhe replied:
In the context of a generic component, TFieldValues is just another generic variable, just as TName. Therefore the TResult type cannot be infered.

The goal of a generic component is to define itself the type it expects, so it can't infer it, it has to define it itself.

kotarella1110 replied:
I see. Thanks!

kotarella1110 commented on src/types/controller.ts:
```diff
@@ -6,9 +6,10 @@ import {
   FieldValues,
+  IsAny,
   RefCallBack,
-  UnpackNestedValue,
```
Why did you remove the `UnpackNestedValue` type?

kotarella1110 replied:
`NestedValue` type is still used for `errors`.

```ts
type FormValues = {
  test: NestedValue<{ custom: string }>;
}
const { fomState: { errors } } = useForm<FormValues>();
```

It is still needed 👍🏻

julienfouilhe replied:
I don't get what it's for, and therefore how to fix it.

Could you maybe provide a test that doesn't pass? 
Why would someone want to use NestedValue in your example. What is the benefit?

julienfouilhe replied:
I added it back on `FieldPathValue`, but I don't know if I should use it for `TResult` too?

kotarella1110 replied:
please see docs: https://react-hook-form.com/ts#NestedValue
The `NestedValue` type is used when the field value is an object.

The following is an example where field name is `test` and field value is `{ custom: string }`.

```ts
type FormValues = {
  test: NestedValue<{ custom: string }>;
}
const { fomState: { errors } } = useForm<FormValues>();
// errors
// {
//    test?: FieldError;
// }
```

If `NestedValue` is not used, it is not possible to derive the correct `errors` type.

```ts
type FormValues = {
  test: { custom: string };
}
const { fomState: { errors } } = useForm<FormValues>();
// errors
// {
//    test?: {
//      custom?: FieldError;
//  }
// }
```

kotarella1110 replied:
> I added it back on `FieldPathValue`

Thanks! looks good!

>  don't know if I should use it for `TResult` too?

This is a tough one...
@bluebill1049 What do you think?

julienfouilhe replied:
@kotarella1110 I'll add tests. It will show if unpacking is needed :)

julienfouilhe replied:
@kotarella1110 Sorry for the delay, I added a test! Definitely needed to use `UnpackNestedValues` for `defaultValue` and `value`. It would not pass otherwise! Thanks for catching that!

kotarella1110 reviewed (commented):
Amazing!!! 🙌
Please check my comments.

kotarella1110 commented on src/types/controller.ts:
```diff
@@ -24,49 +25,69 @@ export type ControllerFieldState<
+  >,
 > = {
-  field: ControllerRenderProps<TFieldValues, TName>;
+  field: ControllerRenderProps<TFieldValues, TResult, TName>;
   formState: UseFormStateReturn<TFieldValues>;
   fieldState: ControllerFieldState<TFieldValues, TName>;
```
Can you fix the `fieldState.error` type in the same way? 🙏🏻 

```ts
export type ControllerFieldState<
  TFieldValues extends FieldValues = FieldValues,
  TResult = any,
  TName extends FieldPathWithValue<TFieldValues, TResult> = FieldPathWithValue<
    TFieldValues,
    TResult
  >,
> = {
  invalid: boolean;
  isTouched: boolean;
  isDirty: boolean;
  error?: FieldErrors<
    IsAny<TResult> extends true ? FieldPathValue<TFieldValues, TName> : TResult
  >;
};
```

julienfouilhe replied:
@kotarella1110 You're right, just did! Thanks!

bluebill1049 replied:
thanks @julienfouilhe 👍

julienfouilhe committed:
✨  Use TResult for `ControllerFieldState` too

julienfouilhe committed:
✨  Put `UnpackNestedValue` back

bluebill1049 committed:
Merge branch 'master' into generic-components

bluebill1049 committed:
Merge branch 'master' into generic-components

bluebill1049 committed:
Merge branch 'master' into generic-components

bluebill1049 committed:
Merge branch 'master' into generic-components

julienfouilhe committed:
✅  Add test for generic components with NestedValue

bluebill1049 committed:
Merge branch 'master' into generic-components

bluebill1049 commented:
@julienfouilhe looks like my last merge from master break your buid. could you take a look at it? 🙏

julienfouilhe committed:
🐛  Fix issues with new fielderrors

julienfouilhe commented:
@bluebill1049 Fixed. Seems like `FieldErrors` behaviour changed and did not like very much primitives anymore.

```typescript
type t = FieldErrors<string>; // fails because string is not compatible with FieldValues instead of returning FieldError
type v = FieldErrors<NestedValue<{ test: string}>>; // { [$NestedValue]?: undefined; test?: FieldError | undefined; } instead of FieldError
```

bluebill1049 commented:
I think that make sense, thanks for the update.

barrymay commented on src/__tests__/useController.test.tsx:
```diff
@@ -33,6 +38,82 @@ describe('useController', () => {
+      return <Generic name="key.0" control={control} />;
+    };
+
+    render(<Component />);
```
thanks for the test, but can you add an expectation of the result?

bluebill1049 commented:
@kotarella1110 would you like to re-review this PR?

kotarella1110 commented on src/__tests__/useController.test.tsx:
```diff
@@ -33,6 +38,82 @@ describe('useController', () => {
+  });
+
+  it('should be able to access values and error in generic components using NestedValue', () => {
+    type ExpectedType = NestedValue<{ test: string }>;
```
Awesome!

kotarella1110 reviewed (dismissed):
LGTM 🎉

julienfouilhe committed:
Merge remote-tracking branch 'origin/master' into generic-components

julienfouilhe committed:
✅  Add expectations

julienfouilhe commented:
@bluebill1049 Done! Sorry for the force-push which dismissed @kotarella1110 review! I usually rebase instead of merge so I was confused for a moment. But all good now!

The diff of the merge commit is weird but I just moved a type I had called `ControllerFieldErrors` and that was below `ControllerFieldState` above it, and renamed it `ControllerFieldError` which was the name used in one of the latest commits. 

I also added expectations in my tests as @barrymay suggested. Since they were tests that just ensured everything was typed correctly, I just put basic expectations.

bluebill1049 committed:
Merge branch 'master' into generic-components

bluebill1049 commented:
Thanks for everyone's contribution and review. Let's do another patch release for all the bug fixes before release this as a new minor update. I will merge that once the patch goes out, there are still few small issues yet to be resolved.

bluebill1049 committed:
Merge branch 'master' into generic-components

bluebill1049 commented:
Just some thoughts, if this is a pattern that we all agree on, perhaps we can introduce something similar to `useFieldArray`? any thoughts.

bluebill1049 committed:
Merge branch 'master' into generic-components

julienfouilhe commented:
@bluebill1049 Definitely something that can be applied to it too!
I don't know how people create generic components. I mainly use Controller thus why I implemented it there, but if others use extensively `useFieldArray` (which I haven't used much for now) it could be useful to make that change there later too!

bluebill1049 committed:
Merge branch 'master' into generic-components

bluebill1049 commented on src/types/controller.ts:
```diff
@@ -1,80 +1,115 @@
-  DeepPartialImpl<UnionLike<T>>,
-  FieldError
->;
+type ControllerFieldError<
```
I didn't pay close enough attention here, @kotarella1110 this was completely changed after the merge.

bluebill1049 replied:
which potentially lead to this: https://github.com/react-hook-form/react-hook-form/issues/6679

julienfouilhe replied:
@bluebill1049 I don't understand the linked issue is talking about versions before this was merged right?

bluebill1049 replied:
![Screen Shot 2021-10-05 at 6 41 06 pm](https://user-images.githubusercontent.com/10513364/135980867-7dc8d3f6-5c74-46a7-a6b0-af1946bfddb2.png)

i am just digging around at the moment, see if i can find a fix for it.

bluebill1049 replied:
` DeepMapImpl<DeepPartialImpl<UnionLike<TFieldValues>>, FieldError>`

vs

`DeepMap<DeepPartial<UnionLike<TFieldValues>>, FieldError>`

bluebill1049 replied:
still try to get my head around it.

bluebill1049 replied:
ok confirmed the latest definitely causing the issue, it could be related to the user's typescript version as well, as i couldn't reproduce the issue at the lib level but codesandbox with CRA, after omit TResult and the issue went away.

bluebill1049 replied:
https://codesandbox.io/s/pedantic-elbakyan-s04xv?file=/src/App.tsx

bluebill1049 commented:
I have to revert this one back guys, i don't have quick solution yet and it probably impact a lot of users.

bluebill1049 commented:
we can come back to this, I am sorry to roll this back but it's better not impact more users.

bluebill1049 commented:
I think we still want this awesome feature.

- [ ] reproduce the issue locally (the one in codesadnbox) https://codesandbox.io/s/pedantic-elbakyan-s04xv?file=/src/App.tsx switch to 17.0.0, it's strange behavior as it wasn't having the issue within all the test local lib has...
- [ ] re-open the PR and publish again

barrymay commented:
> switch to 17.0.0

For clarity, switch RHF to 7.17.0 to see the issue.  As you said, very strange this didn't come up in the test run.  I agree this feature is very much worth trying to get right (and also worth ensuring our test runs can catch this too)

bluebill1049 commented:
`react-hook-form@7.18.0-next.0`
